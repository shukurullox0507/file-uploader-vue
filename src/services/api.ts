import axios from 'axios';

const baseURL = 'http://192.168.100.241:9999/api';

const api = axios.create({
    baseURL,
});

export async function uploadFile(file:File): Promise<any> {
    const formData = new FormData();
    formData.append('tenantId', 'test');
    formData.append('module', 'test');
    formData.append('fileName', 'test');
    formData.append('file', file);
    try {
        return await api.post('/file/upload/public', formData);
    } catch (error:any) {
        throw new Error('Error uploading file: ' + error.message);
    }
}
