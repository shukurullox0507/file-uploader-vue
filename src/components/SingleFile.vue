<template>
  <div class="single">
    <h2>Single File Upload</h2>
    <label for="upload" class="upload-label">
      <div v-if="selectedFile">
        <template v-if="selectedFile?.type.includes('zip')">
          <img src="../assets/folder_zip.png" alt="zip">
        </template>
        <template v-else-if="selectedFile?.type.includes('html')">
          <img src="../assets/html.png" alt="html">
        </template>
        <template v-else-if="selectedFile?.type.includes('pdf')">
          <img src="../assets/pdf.png" alt="pdf">
        </template>
        <template v-else-if="selectedFile?.type.includes('sql')">
          <img src="../assets/sql.png" alt="sql">
        </template>
        <template v-else>
          <img src="../assets/nomalum.png" alt="other">
        </template>
      </div>
      <svg v-else fill="#000000" width="100px" height="100px" viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg">
        <path d="M9 3.793V9H7V3.864L5.914 4.95 4.5 3.536 8.036 0l.707.707.707.707 2.121 2.122-1.414 1.414L9 3.793zM16 11v5H0v-5h2v3h12v-3h2z" fill-rule="evenodd"/>
      </svg>

      <input id="upload" type="file" @change="handleFileUpload">
    </label>
    <div class="fileName">
      <p>{{ selectedFile ? selectedFile.name : 'No file selected' }}</p>
      <button @click="handleUpload">Upload File</button>
    </div>

  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import { uploadFile } from '@/services/api';
import {useToast} from 'vue-toast-notification';

interface CustomFile {
  name: string,
  type: string,
  size: number,
  uploadedFile: File,
}


const selectedFile = ref<CustomFile | null>(null);
const MAX_FILE_SIZE = 20 * 1024 * 1024 // => 20mb
const toast = useToast()
const handleFileUpload = (event: Event) => {
  const file = (event.target as HTMLInputElement)?.files?.[0];
  if (file) {
    if (file.size > MAX_FILE_SIZE) {
      toast.error('File size exceeds the maximum allowed size');
      return;
    }
    toast.success('File uploaded successfully')
    selectedFile.value = {
      name: file.name,
      type: file.type,
      size: file.size,
      uploadedFile:file
    };
  }
};

const handleUpload = async () => {
  if (!selectedFile.value) {
    console.error('No file selected');
    return;
  }

  const formData = new FormData();
  formData.append('tenantId', 'test');
  formData.append('module', 'test');
  formData.append('fileName', 'test');
  formData.append('file', selectedFile.value.uploadedFile); // Append the file here

  try {
    const response = await uploadFile(formData);
    console.log('File uploaded:', response);
  } catch (error:any) {
    console.error(error.message);
  }
};
</script>

<style scoped lang="scss">
.upload-label {
  cursor: pointer;
}

#upload {
  display: none;
}

button {
  margin-left: 20px;
  padding: 5px 16px;
  border: none;
  border-radius: 4px;
  background-color: #007bff;
  color: white;
  cursor: pointer;
}

button:hover {
  background-color: #0056b3;
}

svg {
  width: 100px;
  height: 100px;
}

img {
  width: 100px;
  height: 100px;
}
.fileName{
  display: flex;
}
</style>
