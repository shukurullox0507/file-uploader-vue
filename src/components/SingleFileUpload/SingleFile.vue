<script setup lang="ts">
import { ref } from 'vue';
import { uploadFile } from '@/services/api';
import { useToast } from 'vue-toast-notification';

interface CustomFile {
  name: string;
  type: string;
  size: number;
  uploadedFile: File;
  errorMessage: string | undefined;
}

const selectedFile = ref<CustomFile | null>(null);
const props = withDefaults(
    defineProps<{
      maxSizeFile?: number
    }>(),
    {
      maxSizeFile: 4 * 1024 * 1024
    }
)
const toast = useToast();

const handleFileUpload = (event: Event) => {
  const file = (event.target as HTMLInputElement)?.files?.[0];
  const errorMessage= file.size >= props.maxSizeFile? 'File is too large' :undefined;
  if (file) {
    selectedFile.value = {
      name: file.name,
      type: file.type,
      size: file.size,
      uploadedFile: file,
      errorMessage
    };
  }
};

const handleUpload = async () => {
  if (!selectedFile.value) {
    return;
  }

  if (selectedFile.value.size > props.maxSizeFile) {
    toast.error('File size exceeds the maximum allowed size');
    return;
  }

  const formData = new FormData();
  formData.append('tenantId', 'test');
  formData.append('module', 'test');
  formData.append('fileName', 'test');
  formData.append('file', selectedFile.value.uploadedFile);

  try {
    const response = await uploadFile(formData);
    toast.success("File uploaded successfully");
  } catch (error: any) {
    toast.error(error.message);
  }
};

defineExpose({ handleUpload });
</script>


<template>
  <div class="single">
    <h2>Single File Upload</h2>
    <div class="upload-container">
      <label for="upload" class="upload-label-file">
        <div v-if="selectedFile">
          <img v-if="selectedFile?.type.includes('zip')" src="../../assets/folder_zip.png" alt="zip">
          <img v-else-if="selectedFile?.type.includes('html')" src="../assets/html.png" alt="html">
          <img v-else-if="selectedFile?.type.includes('pdf')" src="../../assets/pdf.png" alt="pdf">
          <img v-else-if="selectedFile?.type.includes('sql')" src="../../assets/sql.png" alt="sql">
          <img v-else src="../../assets/nomalum.png" alt="other">
        </div>
        <svg v-else fill="#000000" width="100px" height="100px" viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg">
          <path d="M9 3.793V9H7V3.864L5.914 4.95 4.5 3.536 8.036 0l.707.707.707.707 2.121 2.122-1.414 1.414L9 3.793zM16 11v5H0v-5h2v3h12v-3h2z" fill-rule="evenodd"/>
        </svg>
        <input id="upload" type="file" @change="handleFileUpload">
      </label>
      <p v-if="!selectedFile" >Tap to choose file</p>
      <p v-if="selectedFile?.errorMessage" data-test-error-message >{{selectedFile?.errorMessage}}</p>
    </div>
    <button v-if="selectedFile" @click="handleUpload" >Upload File</button>
  </div>
</template>

<style scoped lang="scss">
@import './style.scss';
</style>
