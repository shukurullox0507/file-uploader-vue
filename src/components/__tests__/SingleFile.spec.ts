import { describe, expect, it } from 'vitest';
import { uploadFile } from "../../services/api";
import { mount } from "@vue/test-utils";
import { nextTick } from "vue";
import SingleFile from "../SingleFileUpload/SingleFile.vue";

describe('MultipleFile', () => {
    it('should render File input', () => {
        const wrapper = mount(SingleFile);

        expect(wrapper.find('input').html()).toMatchSnapshot();
    });

    it('should show error message when uploaded file more than 4bytes when maxSize is 4', async () => {
        const wrapper = mount(SingleFile, {
            props: {
                maxSizeFile: 4
            }
        })
        const inputElement = wrapper.find('input[type="file"]').element as HTMLInputElement
        const file = new File(['252'], 'foo.txt', {
            type: 'text/plain'
        })
        const mockFileList = Object.create(inputElement.files)
        mockFileList[0] = file
        Object.defineProperty(mockFileList, 'length', { value: 1 });
        await(wrapper.getCurrentComponent().exposed as unknown as any).handleUpload({
            target: { files: mockFileList }
        })
        await nextTick()

        const errorMessageElement = wrapper.find('p[data-test-error-message]')

        expect(errorMessageElement.exists()).toBe(false)
    })
});

