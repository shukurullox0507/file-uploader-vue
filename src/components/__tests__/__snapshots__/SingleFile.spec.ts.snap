// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MultipleFile > should render File input 1`] = `"<input data-v-73c4b518="" id="upload" type="file">"`;
